### DISCUSSION TITLE
Prevent username edit from users

### DISCUSSION THREAD

**Discussion by vkaelin** [zitadel/zitadel#7993] [Q&A]
Hello,

We are using the Zitadel username as a unique identifier for our users in our database. We use a daily sync job to fetch and update Zitadel users inside our own DB.

The issue is that users can edit their username, as shown in the screenshot below, which breaks our identification flow.
<img width="733" alt="Pasted Graphic 3" src="https://github.com/zitadel/zitadel/assets/33511688/59b7dfd5-d8c2-4d35-9632-5ea067216bfc">

We are aware of open issues like #3965 and #6433 but we are looking at solutions in the meantime.

Here are the ideas we have:

1. When a user changes his username, we can see an event is sent (`EventTypes.user.username.changed`). Is it possible to have the old and the new username in this event payload? From what we have seen, we only get the new one. Otherwise, is it possible to cancel/block the event by creating a Zitadel action?

2. In addition to the Zitadel username, we need to store the Zitadel ID of each user too. To be able to know when a user has changed his username and make the necessary changes. But this solution couples our database to Zitadel even more strongly.

What do you think? Do you see a better approach?

Thanks a lot!

**Comment by hifabienne** [comment]
hei @vkaelin 
Generally said, we recommend building your own profile page directly into your applications as this is mostly easier for the end users to understand. this would allow you to remove that button (even if the api would still allow it)

It is not possible to get the old username in the event, as the events are change tracks and only store the difference to the previous state. You could read all the events of the user and compute the state before.
However I also think the approach of sending an error response such as "Changing username is not allowed" would be nice to implement with the actions, and I think it should be possible with our new actions.
You can find the documentation here: https://zitadel.com/docs/apis/actionsv2/introduction

@adlerhurst @stebenz Can you please verify that this would be a usecase possible to solve with the new actions?

**Reply by vkaelin** [reply]
Thanks a lot for your answer, we will check these actions v2!
